datafusion/sql/src/relation/join.rs

```diff
@@ -142,10 +142,29 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                     .build()
             }
             JoinConstraint::Natural => {
-                // https://issues.apache.org/jira/browse/ARROW-10727
-                Err(DataFusionError::NotImplemented(
-                    "NATURAL JOIN is not supported (https://issues.apache.org/jira/browse/ARROW-10727)".to_string(),
-                ))
+                // TODO: what if no common join cols? -> becomes cross join
+                // TODO: limit join_type to specific set?
+                let left_cols: HashSet<&String> = left
+                    .schema()
+                    .fields()
+                    .iter()
+                    .map(|f| f.field().name())
+                    .collect();
+                let keys: Vec<Column> = right
+                    .schema()
+                    .fields()
+                    .iter()
+                    .map(|f| f.field().name())
+                    .filter(|f| left_cols.contains(f))
+                    .map(Column::from_name)
+                    .collect();
```

Comment:
Please notice `qualifier`.

For example,
t1: |t1.a|t1.b|c|   (we can meet `c` instead of `t1.c`)

We also need add test to consider it.

Reply:
I'm not sure I follow, are you suggesting to consider the qualifier when checking for common columns? My understanding is that when joining two different tables they should have different qualifiers anyway.

Reply:
> I'm not sure I follow, are you suggesting to consider the qualifier when checking for common columns? My understanding is that when joining two different tables they should have different qualifiers anyway.

Look like current implementation is ok. `field.name()` don't use qualifier, so this Problem don't exist